File: src/Day5/Boundary.php
<?php

namespace Werner\Adventofcode\Day5;

class Boundary
{
    public function __construct(public int $dest, public int $source, public int $range)
    {
    }

    public static function fromStrings(string $dest, string $source, string $range): self
    {
        return new self(
            dest: (int) $dest,
            source: (int) $source,
            range: (int) $range
        );
    }

    public function hit(int $value): bool
    {
        return $value >= $this->source && $value < $this->source + $this->range;
    }

    public function convert(int $value): int
    {
        if (!$this->hit($value)) {
            return $value;
        }

        return $this->dest - $this->source + $value;
    }
}

File: 2023/5/boundary.go
package main

import (
	"fmt"
)

type Boundary struct {
	Dest, Source, Range int
	conversionOffset    int
	maxSource           int
	Name                string
}

func (b Boundary) Hit(i int) bool {
	return i >= b.Source && i <= b.maxSource
}

func (b Boundary) Convert(i int) int {
	if !b.Hit(i) {
		//fmt.Println("not hit", i, "with", b.String(), "to", i)
		return i
	}

	//fmt.Println("converting", i, "with", b.String(), "to", i+b.conversionOffset)

	return i + b.conversionOffset
}

func (b Boundary) String() string {
	return fmt.Sprintf("%s: [%d-%d] => [%d-%d]", b.Name, b.Source, b.Source+b.Range, b.Dest, b.Dest+b.Range)
}

func NewBoundary(name string, d, s, r int) *Boundary {
	return &Boundary{
		Name:             name,
		Dest:             d,
		Source:           s,
		Range:            r,
		conversionOffset: d - s,
		maxSource:        s + r,
	}
}
